collect_app/src/test/java/org/odk/collect/android/widgets/UrlWidgetTest.java

```diff
@@ -22,14 +33,74 @@
 @RunWith(RobolectricTestRunner.class)
 public class UrlWidgetTest {
 
+    private CustomTabHelper customTabHelper;
+
+    @Before
+    public void setUp() {
+        customTabHelper = mock(CustomTabHelper.class);
+    }
+
+    @Test
+    public void getAnswer_whenPromptAnswerDoesNotHaveAnswer_returnsNull() {
+        assertThat(createWidget(promptWithAnswer(null)).getAnswer(), nullValue());
+    }
+
     @Test
-    public void usingReadOnlyOptionShouldMakeAllClickableElementsDisabled() {
+    public void getAnswer_whenPromptHasAnswer_returnsAnswer() {
+        UrlWidget widget = createWidget(promptWithAnswer(new StringData("blah")));
+        assertThat(widget.getAnswer().getDisplayText(), equalTo("blah"));
+    }
+
+    @Test
+    public void usingReadOnlyOption_makeAllClickableElementsDisabled() {
         UrlWidget widget = createWidget(promptWithReadOnly());
         Button urlButton = widget.findViewById(R.id.url_button);
         assertThat(urlButton.getVisibility(), equalTo(View.GONE));
     }
 
+    @Test
+    public void clearAnswer_doesNotClearWidgetAnswer() {
+        UrlWidget widget = createWidget(promptWithAnswer(new StringData("blah")));
+        widget.clearAnswer();
+        assertThat(widget.getAnswer().getDisplayText(), equalTo("blah"));
+    }
+
+    @Test
+    public void whenPromptHasAnswer_setsCorrectAnswer() {
+        UrlWidget widget = createWidget(promptWithAnswer(new StringData("blah")));
+        assertThat(((TextView) widget.findViewById(R.id.url_answer_text)).getText().toString(), equalTo("blah"));
+    }
+
+    @Test
+    public void whenPromptAnswerDoesNotHaveAnswer_setsNullValue() {
+        UrlWidget widget = createWidget(promptWithAnswer(null));
+        assertThat(((TextView) widget.findViewById(R.id.url_answer_text)).getText().toString(), equalTo(""));
+    }
+
+
+    @Test
+    public void clickingButtonWhenUrlIsEmpty_doesNotCallOpenUri() {
+        UrlWidget widget = createWidget(promptWithAnswer(null));
+        Button urlButton = widget.findViewById(R.id.url_button);
+        urlButton.performClick();
+
+        verify(customTabHelper, never()).bindCustomTabsService(null, null);
+        verify(customTabHelper, never()).openUri(null, null);
+    }
+
+    @Test
+    public void clickingButtonWhenUrlIsNotEmpty_callsCorrectMethods() {
+        UrlWidget widget = createWidget(promptWithAnswer(new StringData("blah")));
+        Button urlButton = widget.findViewById(R.id.url_button);
+        TextView textView = widget.findViewById(R.id.url_answer_text);
+        urlButton.performClick();
+
+        assertThat(textView.getText().toString(), equalTo("blah"));
+        verify(customTabHelper, times(1)).bindCustomTabsService(widget.getContext(), null);
```

Comment:
I'd not bother with `times` here. I'd usually only use it when something happens more than once but I expect a dependency to only get called once in that case (maybe when testing caching or something).

Reply:
Oh okay thanks. I will keep it in mind.